File: android/app/src/main/java/sol/vortexx/android/chat/impl/MessageActionsImpl.kt
package sol.vortexx.android.chat.impl

import io.ktor.client.request.delete
import io.ktor.client.request.post
import io.ktor.client.request.put
import io.ktor.client.request.setBody
import io.ktor.http.ContentType
import io.ktor.http.contentType
import io.ktor.http.isSuccess
import kotlinx.serialization.Serializable
import sol.vortexx.android.chat.api.MessageActions
import sol.vortexx.android.chat.api.MessageSender
import sol.vortexx.android.chat.api.SendOutcome
import sol.vortexx.android.crypto.api.Aead
import sol.vortexx.android.crypto.util.Hex
import sol.vortexx.android.db.dao.MessageDao
import sol.vortexx.android.db.dao.ReactionDao
import sol.vortexx.android.db.entities.ReactionEntity
import sol.vortexx.android.keys.api.KeyAcquisition
import sol.vortexx.android.keys.api.RoomKeyProvider
import sol.vortexx.android.net.impl.VortexHttpClient
import javax.inject.Inject
import javax.inject.Singleton

/**
 * Server-driven message actions. Each method mirrors a `/api/messages/*`
 * endpoint on the node. Local DB updates are optimistic — the server's
 * WS broadcast (`message_updated`, `reaction_added`, …) reconciles the
 * final state in [ChatEngine].
 */
@Singleton
class MessageActionsImpl @Inject constructor(
    private val http: VortexHttpClient,
    private val aead: Aead,
    private val keys: RoomKeyProvider,
    private val messages: MessageDao,
    private val reactions: ReactionDao,
    private val sender: MessageSender,
) : MessageActions {

    override suspend fun react(messageId: Long, emoji: String): Boolean = runCatching {
        // Optimistic — we don't yet know the current user id locally, so
        // the DAO row uses 0 as a placeholder. The WS echo will overwrite.
        reactions.add(ReactionEntity(
            messageId = messageId, userId = 0, emoji = emoji,
            createdAt = System.currentTimeMillis(),
        ))
        val resp = http.client.post("api/messages/$messageId/react") {
            contentType(ContentType.Application.Json); setBody(ReactReq(emoji))
        }
        resp.status.isSuccess()
    }.getOrDefault(false)

    override suspend fun edit(messageId: Long, newPlaintext: String): Boolean {
        val existing = messages.byId(messageId) ?: return false
        val keyAcq = keys.keyFor(existing.roomId)
        if (keyAcq !is KeyAcquisition.Ready) return false
        val newCt = Hex.encode(aead.encrypt(Hex.decode(keyAcq.keyHex), newPlaintext.toByteArray()))

        messages.upsert(existing.copy(
            plaintext = newPlaintext, ciphertextHex = newCt,
            editedAt = System.currentTimeMillis(),
        ))
        return runCatching {
            http.client.put("api/messages/$messageId") {
                contentType(ContentType.Application.Json)
                setBody(EditReq(newCt))
            }.status.isSuccess()
        }.getOrDefault(false)
    }

    override suspend fun reply(roomId: Long, replyToMessageId: Long, plaintext: String): SendOutcome {
        // Server-side reply is just a normal send with reply_to field; for
        // the client, we queue via MessageSender and tag the local row.
        val outcome = sender.send(roomId, plaintext)
        if (outcome is SendOutcome.Queued) {
            messages.lastN(roomId, 1).firstOrNull()?.let {
                messages.upsert(it.copy(replyTo = replyToMessageId))
            }
        }
        return outcome
    }

    override suspend fun delete(messageId: Long): Boolean = runCatching {
        messages.delete(messageId)
        http.client.delete("api/messages/$messageId").status.isSuccess()
    }.getOrDefault(false)

    override suspend fun openThread(messageId: Long): Long {
        // Thread id = deterministic mapping over messageId so the UI
        // router can open "thread:MSG" without an extra round-trip. Server
        // creates the virtual room on first send.
        return -messageId   // negative id convention distinguishes from real rooms
    }

    @Serializable private data class ReactReq(val emoji: String)
    @Serializable private data class EditReq(val ciphertext: String)
}

File: ios/Modules/Sources/Chat/impl/MessageActionsImpl.swift
import Foundation
import Net
import VortexCrypto
import DB
import Keys
import Threads

/// Server-driven message actions. Each method mirrors a `/api/messages/*`
/// endpoint on the node. Local DB updates are optimistic — the server's
/// WS broadcast reconciles the final state in [ChatEngine].
public final class MessageActionsImpl: MessageActions, @unchecked Sendable {
    private let http: HttpClient
    private let aead: Aead
    private let keys: RoomKeyProvider
    private let messages: MessageDao
    private let reactions: ReactionDao
    private let sender: MessageSender
    private let threads: ThreadsRepository

    public init(
        http: HttpClient,
        aead: Aead,
        keys: RoomKeyProvider,
        messages: MessageDao,
        reactions: ReactionDao,
        sender: MessageSender,
        threads: ThreadsRepository,
    ) {
        self.http = http
        self.aead = aead
        self.keys = keys
        self.messages = messages
        self.reactions = reactions
        self.sender = sender
        self.threads = threads
    }

    public func react(messageId: Int64, emoji: String) async -> Bool {
        do {
            // Optimistic — userId = 0 placeholder, server echo overwrites
            // with the real sender_pseudo.
            try await reactions.upsert(ReactionRecord(
                messageId: messageId, userId: 0, emoji: emoji,
                createdAt: Int64(Date().timeIntervalSince1970 * 1000)))
            let req = try HttpRequest.postJson(
                "api/messages/\(messageId)/react",
                body: ReactReq(emoji: emoji))
            _ = try await http.send(req)
            return true
        } catch { return false }
    }

    public func edit(messageId: Int64, newPlaintext: String) async -> Bool {
        guard let existing = try? await messages.byId(messageId) else { return false }
        guard case .ready(let keyHex, _) = await keys.keyFor(existing.roomId) else { return false }
        do {
            let key = try Hex.decode(keyHex)
            let ct  = try aead.encrypt(key: key, plaintext: Data(newPlaintext.utf8))
            let newCtHex = Hex.encode(ct)
            var row = existing
            row.plaintext = newPlaintext
            row.ciphertextHex = newCtHex
            row.editedAt = Int64(Date().timeIntervalSince1970 * 1000)
            try await messages.upsert(row)

            let req = HttpRequest(
                method: .PUT,
                path: "api/messages/\(messageId)",
                body: try JSONEncoder().encode(EditReq(ciphertext: newCtHex)),
                extraHeaders: ["Content-Type": "application/json"],
            )
            _ = try await http.send(req)
            return true
        } catch { return false }
    }

    public func reply(roomId: Int64, replyToMessageId: Int64, plaintext: String) async -> SendOutcome {
        let outcome = await sender.send(roomId: roomId, plaintext: plaintext)
        if case .queued(let localId) = outcome,
           let row = try? await messages.byId(localId) {
            var edited = row
            edited.replyTo = replyToMessageId
            try? await messages.upsert(edited)
        }
        return outcome
    }

    public func delete(messageId: Int64) async -> Bool {
        do {
            try await messages.delete(messageId)
            _ = try await http.send(.delete("api/messages/\(messageId)"))
            return true
        } catch { return false }
    }

    /// Creates a thread rooted at `messageId` via the server's
    /// `/api/rooms/{roomId}/threads` endpoint. Resolves the parent's
    /// roomId from the local cache — both the ECIES flow and the
    /// public-key flow guarantee the message row is already there
    /// before the user can tap "Open thread".
    public func openThread(messageId: Int64, title: String?) async -> Threads.Thread? {
        guard let parent = try? await messages.byId(messageId) else { return nil }
        let headline = title
            ?? parent.plaintext.map { String($0.prefix(40)) }
            ?? "Thread"
        return await threads.create(
            roomId: parent.roomId,
            parentMessageId: messageId,
            title: headline,
        )
    }

    private struct ReactReq: Encodable { let emoji: String }
    private struct EditReq: Encodable  { let ciphertext: String }
}
